core/trino-main/src/main/java/io/trino/operator/ExchangeOperator.java

```diff
@@ -99,16 +105,24 @@ public SourceOperator createOperator(DriverContext driverContext)
                         retryPolicy,
                         exchangeManagerRegistry);
             }
-            return new ExchangeOperator(
+            ExchangeOperator exchangeOperator = new ExchangeOperator(
                     operatorContext,
                     sourceId,
                     exchangeDataSource,
-                    serdeFactory.createPagesSerde());
+                    serdeFactory.createPagesSerde(), noMoreSplitsTracker);
+            noMoreSplitsTracker.operatorAdded(exchangeOperator);
+            return exchangeOperator;
         }
 
         @Override
         public void noMoreOperators()
         {
+            noMoreSplitsTracker.noMoreOperators();
+            if (noMoreSplitsTracker.isNoMoreSplits()) {
+                if (exchangeDataSource != null) {
+                    exchangeDataSource.noMoreInputs();
```

Comment:
this can be called multiple times due to races. Is that a problem?

Reply:
This is allowed in other place, see https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/server/protocol/Query.java#L587 which can be called multiple times today through https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java#L187. Though I don't have a strong opinion here.